openshift/CLOUD_3928_Custom_galleon_layers.adoc

```diff
@@ -0,0 +1,113 @@
+= Evolve support for custom Galleon feature-packs usage during s2i.
+:author:           Jean-Francois Denise
+:email:             jdenise@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Today the ability to integrate custom Galleon feature-packs and layers during s2i build phase is limited and not flexible. We are here proposing
+to evolve it with new configuration points and behavior.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/CLOUD-3928[CLOUD-3928]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-1647[EAP7-1647]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:TBD
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+* [ ] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+* https://github.com/wildfly/wildfly-s2i/[WildFly S2i]
+
+* https://github.com/wildfly/wildfly-cekit-modules/[WildFly cekit modules]
+
+
+=== Other Interested Projects
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [ ] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [ ] Managed domain
+
+* [x] OpenShift s2i
+
+* [ ] Bootable jar
+
+== Requirements
+
+* Ability to configure the location of the custom Galleon content a user is providing.
+
+* Ability to provide a settings.xml file used during galleon provisioning of the server.
+
+* Ability to locate un-released custom Galleon feature-packs in the application project. Custom galleon feature-packs are then copied to local maven cache 
+to be used during provisioning.
+
+* Ability to provide a list of custom Galleon feature-packs without requiring a provisioning.xml file. 
+
+=== Hard Requirements
+
+* Define new env variable: `GALLEON_DIR=<path relative to app root dir that contains Galleon content>`. Default being `galleon`. If the referenced directory doesn't
+exist, the build phase aborts.
+
+ * A settings.xml file can be located in the `${GALLEON_DIR: galleon}` dir and used during provisioning. Any local repository configured in such settings are ignored. 
+If no settings.xml is found, a default settings.xml is used.
+
+* Define a new env variable: `GALLEON_CUSTOM_FEATURE_PACKS_MAVEN_REPO=<absolute path to a maven repo dir containing custom galleon feature-packs>`
+By default we are looking into the `<src root dir>/${GALLEON_DIR:galleon}/repository`. Any path can be set for `GALLEON_CUSTOM_FEATURE_PACKS_MAVEN_REPO`, this allows to mount a volume containing the maven repo.
+During s2i build, the content of the maven repository is copied to the local maven repository present in the image used to provision a server.
```

Comment:
Does that mean `copied over`, ie. GALLEON_CUSTOM_FEATURE_PACKS_MAVEN_REPO can be used to override galleon repo available in builder image?

Reply:
No, this repo is only for custom feature-packs, the local repo can't be set.

Reply:
so in maven terminology this repo will serve as another remote repo for the build? eg. something like:
```xml
<profile>
      <id>galleon-custom-fp-repo-profile</id>
      <repositories>
        <repository>
          <id>galleon-custom-fp-repo</id>
          <url>file://${env.GALLEON_CUSTOM_FEATURE_PACKS_MAVEN_REPO}</url>
          <releases>
            <enabled>true</enabled>
          </releases>
          <snapshots>
            <enabled>true</enabled>
          </snapshots>
        </repository>
      </repositories>
</profile>
```

Reply:
We keep the builder image local repo, we simply copy into it the custom repository content.

Reply:
That's what I asked originally, so I suppose that the answer there should have been yes.
In other words, is the content of `GALLEON_CUSTOM_FEATURE_PACKS_MAVEN_REPO` somehow restricted or it can contain any maven artifacts? Let's suppose I would like to use it to provide all artifacts needed to build an app (not just FPs) to perform offline build - is that supported scenario?

Reply:
It is unsupported scenario, we will only document that this is the place where to put Galleon related artifacts, not application artifacts.

Reply:
Thanks.

Btw, what about shorter name of the variable? eg. just `GALLEON_MAVEN_REPO` or `GALLEON_CUSTOM_MAVEN_REPO`?

Reply:
I agree that the name is long but it also makes it clear that the repo is supposed to contain only custom galleon feature-packs.

Reply:
ok, that was last question, thanks for your patience.